### PR TITLE
SR-IOV: MSI/MSI-X interrupt new TC

### PR DESCRIPTION
**Pull request by adriansuhov**
New TC "SR-IOV Interrupts change": verifies if the interrupts count from /proc/interrupts is increasing after traffic goes through the VF.
Expected initial results:
a. The number of PCIe MSI devices processing interrupts should be greater than NUM_VF*9

Steps:
1. On the controller-vm save/construct these values from /proc/interrupts:
    a. A sum of all interrupts per PCIe MSI device (exclude MSI devices with 0 interrupts)
    b. A list of all vCPUs that are processing interrupts. Save the number of interrupts for each vCPU.
2. Start iPerf3 for 10 minutes with 128 threads
3. Repeat step 1.a & 1.b after iPerf3 finished

Expected results:
After comparing the values from step 1 and step 3:
- each value from step 3.a should increase after the iPerf3 run.
- at least half of the values from step 3.b should increase.
Note: I'm using 'at least half of the values' as reference because in all of my tests there are a few vCPUs that doesn't increase the interrupts count. This is also platform & distro dependent and is necessary to avoid false positives. 

# Test results
[LISAv2 Test Results Summary]
Test Run On           : 03/14/2019 22:36:49
ARM Image Under Test  : **Canonical : UbuntuServer : 18.04-LTS : latest**
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:22

   ID TestCaseName                                                                TestResult TestDuration(in minutes)
---------------------------------------------------------------------------------------------------------------------
    1 SRIOV-INTERRUPTS-CHANGE                                                           PASS                14.33

[LISAv2 Test Results Summary]
Test Run On           : 03/14/2019 22:38:32
ARM Image Under Test  : **RedHat : RHEL : 7.5 : latest**
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:25

   ID TestCaseName                                                                TestResult TestDuration(in minutes)
---------------------------------------------------------------------------------------------------------------------
    1 SRIOV-INTERRUPTS-CHANGE                                                           PASS                16.75

### CODE REVIEW

File: XML/TestCases/FunctionalTests-SRIOV.xml [hunk 1]
```diff
@@ -122,6 +122,35 @@
         <Tags>sriov,network</Tags>
         <Priority>0</Priority>
     </test>
+    <test>
+        <testName>SRIOV-INTERRUPTS-CHANGE</testName>
+        <setupScript>.\TestScripts\Windows\SETUP-Change-CPU.ps1,.\Testscripts\Windows\SETUP-SR-IOV-Configure.ps1</setupScript>
```
Comment (LiliDeng) [inline XML/TestCases/FunctionalTests-SRIOV.xml:127]: This only can be apply in hyperv, this case belongs azure platform

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+# Get initial values of PCIe MSI devices processing interrupts and
+# values for each vCPU
+get_interrupts_per_pci "1"
+# Compute an array with interrupts/vCPU before the traffic starts
+get_interrupts_per_vcpu "1"
+
+# Start iPerf3 to force the interrupt count to increase
+LogMsg "------------ Running iPerf3 for 10 minutes ------------"
+ssh -i "$HOME"/.ssh/"$SSH_PRIVATE_KEY" -o StrictHostKeyChecking=no "$remote_user"@"$VF_IP2" "nohup iperf3 -s > /dev/null 2>&1 &"
```
Comment (LiliDeng) [inline Testscripts/Linux/SR-IOV_Interrupts_Change.sh:123]: Should we check if iperf3 exists or not?

File: XML/TestCases/FunctionalTests-SRIOV.xml (hunk 1 above)
Comment (juhlee-microsoft) [reply XML/TestCases/FunctionalTests-SRIOV.xml:127]: We knew this and would like to scale out to Azure.

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh (hunk 2 above)
Comment (adriansuhov) [reply Testscripts/Linux/SR-IOV_Interrupts_Change.sh:123]: It is always installed here: https://github.com/LIS/LISAv2/blob/master/Testscripts/Windows/SRIOV-RUN-COMMON-TEST.ps1#L112

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        exit 0
+    fi
+done
+
+cpu_len=${#initial_cpu_interrupts[@]}
+for (( cpu=0; cpu<$cpu_len; cpu++ ))
+do
+    if [[ "${initial_cpu_interrupts[$cpu]}" -ge "${final_cpu_interrupts[$cpu]}" ]] && \
+       [[ "${initial_cpu_interrupts[$cpu]}" -gt "1000" ]]; then
+        LogErr "At least one vCPU didn't have an increased interrupt count after iPerf3 run!"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SR-IOV_Interrupts_Change.sh]: At least, no vCPU has an interrupt count increment after iperf3 running

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+get_interrupts_per_pci "2"
+# Compute an array with interrupts/vCPU after the traffic stopped
+get_interrupts_per_vcpu "2"
+
+# Compare the final arrays with the initial arrays
+pci_len=${#initial_pci_interrupts[@]}
+for (( pci=0; pci<$pci_len; pci++ ))
+do
+    if [ "${initial_pci_interrupts[$pci]}" -ge "${final_pci_interrupts[$pci]}" ]; then
+        LogErr "At least one MSI PCIe device didn't have an increased interrupt count after iPerf3 run!"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SR-IOV_Interrupts_Change.sh]: one MSI PCIe device -> one PCI device

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

**Comment by juhlee-microsoft** [comment]
Can test log includes before/after interrupt count snapshot as raw data?

File: XML/TestCases/FunctionalTests-SRIOV.xml (hunk 1 above)
Comment (adriansuhov) [reply XML/TestCases/FunctionalTests-SRIOV.xml:127]: Actually I forgot to put the Hyper-V tag. I will put test results for Hyper-V too

**Comment by adriansuhov** [comment]
@juhlee-microsoft They include the interrupts/vcpu before & after the test run. Here is an example:
[lisa@LISAv2~]$ cat TestExecution.log
Thu Mar 14 23:16:13 2019 : Testscript running on redhat_7
Thu Mar 14 23:16:13 2019 : Successfully initialized testscript!
Thu Mar 14 23:16:31 2019 : yum_install ruby: Success
Thu Mar 14 23:16:31 2019 : Expected VF count: 1. Actual VF count: 1
Thu Mar 14 23:16:31 2019 : The number of MSI PCIe devices processing interrupts is: 17
Thu Mar 14 23:16:31 2019 : vCPU 0 is processing 1157 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 1 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 2 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 3 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 4 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 5 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 6 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 7 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 8 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 9 is processing 0 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 10 is processing 3907 interrupts
Thu Mar 14 23:16:31 2019 : vCPU 11 is processing 0 interrupts
------ [adsuho]: skipping logs ------
Thu Mar 14 23:16:33 2019 : ------------ Running iPerf3 for 10 minutes ------------
Thu Mar 14 23:26:35 2019 : The number of MSI PCIe devices processing interrupts is: 17
Thu Mar 14 23:26:35 2019 : vCPU 0 is processing 1157 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 1 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 2 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 3 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 4 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 5 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 6 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 7 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 8 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 9 is processing 0 interrupts
Thu Mar 14 23:26:35 2019 : vCPU 10 is processing 1822779 interrupts
------ [adsuho]: skipping logs ------

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh (hunk 4 above)
Comment (adriansuhov) [reply Testscripts/Linux/SR-IOV_Interrupts_Change.sh]: Will change

**Comment by adriansuhov** [comment]
I've further consolidated the test script & also did a test run on Hyper-V:
[LISAv2 Test Results Summary]
Test Run On           : 03/15/2019 08:20:59
VHD Under Test        : ****\Ubuntu\ubuntu_18.04.1\ubuntu_18.04.1_gen1_gen2.vhdx
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:24

   ID TestCaseName                                                                TestResult TestDuration(in minutes)
---------------------------------------------------------------------------------------------------------------------
    1 SRIOV-INTERRUPTS-CHANGE                                                           PASS                18.84


Improved the logging:
Fri Mar 15 09:02:44 2019 : vCPU 5 interrupts increased from 166 to 925643
Fri Mar 15 09:02:44 2019 : vCPU 7 interrupts increased from 159 to 937679
Fri Mar 15 09:02:44 2019 : vCPU 9 interrupts increased from 997 to 1674760
Fri Mar 15 09:02:44 2019 : vCPU 14 interrupts increased from 351 to 871065
Fri Mar 15 09:02:44 2019 : vCPU 16 interrupts increased from 592 to 1189431
Fri Mar 15 09:02:44 2019 : vCPU 18 interrupts increased from 219 to 2793269
Fri Mar 15 09:02:44 2019 : vCPU 21 interrupts increased from 299 to 3123559
Fri Mar 15 09:02:44 2019 : vCPU 34 interrupts increased from 148 to 2053292
Fri Mar 15 09:02:44 2019 : vCPU 38 interrupts increased from 326 to 4770674
Fri Mar 15 09:02:44 2019 : vCPU 44 interrupts increased from 1210 to 1287228
Fri Mar 15 09:02:44 2019 : vCPU 46 interrupts increased from 3644 to 10913
Fri Mar 15 09:02:44 2019 : vCPU 48 interrupts increased from 460 to 3089460
Fri Mar 15 09:02:44 2019 : vCPU 52 interrupts increased from 472 to 2576655
Fri Mar 15 09:02:44 2019 : vCPU 54 interrupts increased from 435 to 1614233
Fri Mar 15 09:02:44 2019 : vCPU 56 interrupts increased from 656 to 3614524
Fri Mar 15 09:02:44 2019 : vCPU 58 interrupts increased from 127 to 3105506
Fri Mar 15 09:02:44 2019 : vCPU 62 interrupts increased from 210 to 1584688

Removing PR-UNDER-TEST & DO-NOT-MERGE labels

File: Testscripts/Linux/SR-IOV_Interrupts_Change.sh [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+. SR-IOV-Utils.sh || {
+    echo "ERROR: unable to source SR-IOV_Utils.sh!"
+    echo "TestAborted" > state.txt
+    exit 0
+}
+UtilsInit
+
+# Install required package
+update_repos
+install_package ruby
```
Comment (chvalean) [inline Testscripts/Linux/SR-IOV_Interrupts_Change.sh]: where are you using ruby?

**Review by juhlee-microsoft — APPROVED** [review]

**Pull request merged**